### DISCUSSION TITLE
Possibility to add Modbus TCP communication

### DISCUSSION THREAD

**Discussion by mfaust78** [mletenay/home-assistant-goodwe-inverter#65] [Ideas]
I was wondering if it would be desirable to add communication via Modbus TCP to inverters.
AFAIK the ET supports Modbus/TCP with the LAN adapter and any inverter with RS-485 port can be connected with a Modbus RTU/TCP gateway.
I've added a Modbus RTU/TCP to my Inverter and can read the same registers now via Modbus TCP on port 502.
Might also be an option to add GoodWe Inverters which do not support/use port 8899 ot Home Assistant.
I would be happy to be testing site for that.

**Comment by mletenay** [comment]
To be honest, I never heard of that TCP gateway, could be interesting.
I guess it would be very welcome to people with inverters types (or old firmwares) which do not talk UDP.
But since I have to direct access to it, it would be rather pain to develop it "in the dark".
Is the TCP gateway goodwe specific or it is simple 3rd party gadget ?
Where to get it or how to build it ?

**Reply by mfaust78** [reply]
I spent some time to find a Modbus RTU/TCP gateway. After searching a while I noticed it is quite common for machine automation. Unfortunately most have steep price tags (>USD 500), but two have reasonable price tags:
https://www.icpdas.com/en/product/tGW-715
https://www.planet.com.tw/en/product/mg-110
The latter (MG-110) does only support IDs from 1 to 32 (247 can be mapped though), the tGW-715 works with IDs from 1-247. Price range is around USD 120.
Both (and others) support multiple RS-485 devices as long as they are at the same baud rate and on the same RS-485 cable. Unfortunately the ET uses 9600 baud for RS-485. Connection is made via the EMS port (RJ45 beside the meter connector).

GoodWe told me that the "ET" supports Modbus TCP natively when the LAN Module is used. Unfortunately this requires removing the WiFi module and USB port which breaks the PV Master app) and therefore is not favorable.

> But since I have to direct access to it, it would be rather pain to develop it "in the dark".

Fully agree. I'm not very good at developing, but could try.

> Is the TCP gateway goodwe specific or it is simple 3rd party gadget ?

Not GoodWe specific at all. Different 3rd party suppliers offer Gateways. The only GoodWe specific thing are the Modbus ID and the Modbus Registers (but they stay the same from UDP to Modbus RTU to Modbus TCP)

> Where to get it or how to build it ?

Depends on where you live. US should be simple, Europe is a bit tricker.
If you want to build it, there are RS-485 hats for Arduino or Raspberry plus open source code that can be used...

**Comment by mletenay** [comment]
Well, most of users here and in discord seem to be from Europe, Australia or South Africa and brief googling did not lead to any easy source of those gadgets.
But thinking about it, it smells lot a like some guys are developing here - https://github.com/jantenhove/GoodWeLogger

**Reply by mfaust78** [reply]
I'm in Europe too, got mine from https://www.icp-deutschland.de/en/home/ (tGW-715) and https://www.ersazza.de/index.php?cl=search&searchparam=Planet+MG-110+&lang=1. Not sure about other regions. Anyway there any other Gateways which might be available, but they are somewhat hard to google for.

I had a look at your code (mainly protocol and modbus). Unfortunately it is beyond my skills to adapt it, but I guess the main thing to change is to adapt it to speak TCP instead of UDP and adjust the header/CRC, as the commands and data structure is identical. 

Here an example for reading the first 4 registers from 35000:
Sent: _00 01 00 00 00 06_ F7 03 88 B8 00 04
Rec: _00 01 00 00 00 19_ F7 03 08 00 01 27 10 00 FE 39 30 
The header (in italic) contains:
- two bytes are the 'transaction ID'
- two bytes 0 for Modbus/TCP
- two bytes number of remaining bytes

(source: [Modbus on Wiki](https://en.wikipedia.org/wiki/Modbus#Modbus_TCP_frame_format_(primarily_used_on_Ethernet_networks))) 
There is no CRC at the end compared to RTU (or the RTU/UDP that GoodWe uses).
Beside that the Data is identical...

> But thinking about it, it smells lot a like some guys are developing here - https://github.com/jantenhove/GoodWeLogger

It has a relation, but not the same. If I understood it correctly, they connect directly to RS-485 and then provide the data via MQTT. Also a gateway, but changing the data structure.

**Comment by BenBarat** [comment]
This appears to be an opportunity to sell a lot of 'specialist' RS485 MODBUS converter hardware, when in actual fact all that is needed is a standard $10 Chinese model. Documentation on where and how to physically connect the adapters are in the GoodWe installation manuals available on their website, which it appears many people have overlooked.

The software on the PC/computer emulates a standard serial port to talk to the MODBUS interface. Make sure you match your MODBUS port on your software so both ends can actually talk to each other. This is separate from your serial port configuration (baud rate, stop bits, etc).

The adapters usually have built in opto isolators and transient suppressors so no need to worry about frying your network equipment by sudden impulses. That is why the RS485 is used in so many industrial situations where the electrical environment is very harsh, and the signal has to travel long distances (up to 1.2 kilometers).

Don't forget your terminating resistors need to be on both ends of the RS485 connections to prevent noise ringing of the signal. If you have more than two devices on your MODBUS network (the main reason people go MODBUS and daisy chain them, using the unique port number of each device to individually address them), then only the ones on each end need to be terminated.

Try eBay if you are unsure, or if you are in Australia, Altronics have one for around $90 which seems a lot for a simple interface, but that one is very versatile, and probably overkill. See https://www.altronics.com.au/p/d4231-serial-to-ethernet-converter-rs232-422-485-modbus/ or if you understand that RS485 supports very long cable lengths (over a kilometer), use the https://www.altronics.com.au/p/d4231-serial-to-ethernet-converter-rs232-422-485-modbus/ (also found for around $7 on eBay with free postage) at your PC and chop off both ends of the cable to the inverter using the very same twisted pair you would use for Ethernet by adapting it for RS485 use, bypassing your network altogether.

All the interfaces that GoodWe supplies are actually all emulating a serial port also, whether by TCP or UDP or MODBUS. Once you understand this, it all falls into place. Whether the physical connection is hard wired or across a local network or the internet, to the the higher software level it all appears as a serial connection. Both ends chat to each other using pre-defined codes to send the appropriate queries and appropriate responses.

Death warning: Make sure all your equipment is fully disconnected, and no electricity is flowing. Check with a multimeter to make sure. Work with one hand in your pocket if possible so no possible way an electrical path across your heart will happen should you accidentally plug something in and everything springs into life. Remember that inverter is designed to deliver enough current to power your household, so imagine what it will do to your body. The smell of sizzled flesh when the coroner scrapes your body off the ceiling is not fun. You have been warned!!!

I question why something that works needs to be fixed? If the SEMS portal is being updated then your equipment is fully functional - both ends are actually talking to each other - just send the right packets across your network and get the correct data back. Interpreting it is the tricky question. GoodWe have released a number of papers on their website - just a matter of correctly decoding it.

Good luck!

**Reply by mgx0** [reply]
necropost!

"I question why something that works needs to be fixed?" : you know why? this question is an answer: why the installer or vendor shall be able to push settings to my inverter without my approval, or even my knowledge?

i would like to remove the wifi module and cut my inverter off the internet.

**Reply by janchlebek** [reply]
Removing WiFi? Why don't you simply block internet communication for the IP of the inverter?

**Reply by mgx0** [reply]
> Removing WiFi? Why don't you simply block internet communication for the IP of the inverter?

Because:

1. it still transmits the AP with the default password. It is resetting the SSID back to Solar-XXXX and password to 12345678, which I believe only an ignorant to be willing to have at home. Wandering around the houses and turning the inverters of the people having Goodwe could be a lot of fun though.
2. i want to pull data from the inverter for my local processing. 

And because rs485 does not work with two masters on the bus. I have to at least power off the wifi module in order to stop it from requesting registers from the inverter, so i can connect my own bridge. Am i wrong?

for now i have managed to stop receiving commands from Goodwe and still being able to upload my values to sems. But i still would like to connect directly to the inverter

**Comment by davefx** [comment]
If you want to avoid someone from Goodwe's cloud messing with your
inverter, you can simply assign it an only-local IP address, without
Internet access.
That way your inverter will be accessible from the local network, but not
from Goodwe's cloud.
--
David Marín Carreño ***@***.***>

**Reply by mgx0** [reply]
> If you want to avoid someone from Goodwe's cloud messing with your inverter, you can simply assign it an only-local IP address, without Internet access. That way your inverter will be accessible from the local network, but not from Goodwe's cloud. 

it already has a "local-only" IP, but i know what you mean (removing the gateway so it cannot go through). I have made a solution for myself: i am able to send data to SEMS and my module will not listen to commands from SEMS using a python MITM proxy. But since it does not receive anything from SEMS, the Wifi LED keeps blinking, however i see the data is being sent and tracked in SEMS portal and app.

Who wants details, i can provide it, it's quite easy, but i will not mess up this Issue.

**Reply by BenBarat** [reply]
Is your inverter waiting, breathlessly, for some acknowledgement from SEMS that it successfully received the data, before continuing on with doing whatever else it needs to do? Intercepting the data in your integration by diverting it, in a one way direction, may be the crux of the matter, the inverter chip putting unacknowledged messages on the stack, and waiting to receive acknowledgements before removing them? Soon the stack overflows, the chip resets (hopefully), and the process repeats, much to your chagrin, and that flashing network LED...
I'm guessing here, but knowing how a lot of firmware is lashed together, quite feasible.
You may be trying to work around a fundamental design issue flaw, something I would like to see GoodWe step up to, release more detailed information, open up the APIs, and maybe work with some of the developers here to solve this vexing issue.

**Reply by tabnul** [reply]
old message but still i am replying. It seems SEMS / Goodwe has messed with the invertor parameters in recent months/years.

I had a firewall rule blocking internet for the invertor , and accessing it locally via modbus. Worked for 2 years.
2 months ago i messed up my firewall rule and it had internet access for a while. Apparently it got settings pushed from SEMS.

Now, after re-implementing the firewall rule, and flashing the exact same ARM and DSP firmware, i cannot get it to work while the internet is blocked. Well, very shortly it seems to work after a restart.

Instant fix: allow internet for the stick.
Re-deploy firewall rule: keeps working!
Next day or after a full DC power down: not working anymore.

Apparently a succesfull connection with SEMS on 20001 opens up the modbus local service untill a restart happens .

Not sure what to do about this, ordered an old LAN stick, hoping that will help. If not , will go full DIY.

For the record:
gw6k-dt
ARM + DSP v16 (tried upgrading to 23/24 )
Wifi/LAN stick
